### PR TITLE
feat(intelligence): add empirical skill creation and comparative anal…

### PR DESCRIPTION
**Pull request by athola**
…ytics

Add three new intelligent skill modules for session-driven insights:

- Empirical skill creation (`--method empirical`): mines Claude Code and Codex CLI session history to extract successful tool sequences and failure anti-patterns, clustering them into reusable skill templates
- Comparative recommendations: deviation scoring compares actual vs expected outcomes per skill category (Testing, Debugging, Documentation) to identify underperforming skills
- Behavioral analytics: extracts tool call sequences, file access patterns, and session outcomes (success/failure/partial) from history

Also includes:
- Install script improvements with user-scoped MCP registration
- Version bump to 0.4.4 across all crates
- Changelog updates with release dates for 0.4.3 and 0.4.4

### CODE REVIEW

**Comment by athola** [comment]
## Test Plan for PR #72

Generated from `/pr-review` on 2026-01-02

### Prerequisites
- [ ] All blocking issues (B1) have been addressed
- [ ] All in-scope issues (S1-S6) have been addressed
- [ ] Code compiles without errors

---

### Blocking Issues Verification

#### B1: Edge case in improvement_factor calculation
**File:** `crates/intelligence/src/recommend/comparative.rs:571`
**Issue:** Division produces 0.0 when without_retry=0.0 and with_retry>0.0

**Verification Steps:**
1. [ ] Review the fix at `crates/intelligence/src/recommend/comparative.rs:571`
2. [ ] Run: `cargo test --package skrills-intelligence -- comparative`
3. [ ] Add test case: baseline with 0.0 retry rate, skill with >0 retry rate
4. [ ] Verify improvement factor handles edge case correctly

**Expected Outcome:**
- Tests pass including new edge case test
- improvement_factor produces sensible value (not 0.0) when baseline has no retries

---

### In-Scope Issues Verification

#### S4: String truncation not Unicode-safe (HIGH PRIORITY)
**File:** `crates/intelligence/src/usage/behavioral.rs:176`
**Issue:** `s[..197]` can panic on UTF-8 boundary

**Verification Steps:**
1. [ ] Review the fix at `crates/intelligence/src/usage/behavioral.rs:176`
2. [ ] Also check lines 213-215, 264-268 for similar patterns
3. [ ] Run: `cargo test --package skrills-intelligence -- behavioral`
4. [ ] Add test case with non-ASCII content exceeding 200 chars

**Expected Outcome:**
- No panics with Unicode input
- Strings truncate cleanly at valid UTF-8 boundaries

#### S6: Empirical method incomplete
**File:** `crates/server/src/app/intelligence.rs:624-699`
**Issue:** Returns preview message instead of generating skills

**Verification Steps:**
1. [ ] Review create_skill_tool implementation
2. [ ] Verify documentation states this is preview-only OR complete implementation
3. [ ] If completing: add integration test for `--method empirical`

**Expected Outcome:**
- Documentation matches implementation behavior
- Clear UX messaging for preview features

---

### Build & Quality Gates

**Run these commands to verify overall quality:**

```bash
# Full test suite
cargo test --all

# Linting
cargo clippy --all -- -D warnings

# Formatting
cargo fmt --all --check

# Build release
cargo build --release
```

**All must pass before PR approval.**

---

### Summary Checklist

| Issue ID | File | Verified | Notes |
|----------|------|----------|-------|
| B1 | comparative.rs:571 | [ ] | Edge case in improvement_factor |
| S4 | behavioral.rs:176,213,264 | [ ] | Unicode-safe truncation |
| S6 | intelligence.rs:624 | [ ] | Empirical preview/complete |
| Build | All | [ ] | `cargo test && cargo clippy && cargo build --release` |

**Ready for merge when all boxes checked.**

---
*Test plan generated by /pr-review - execute with /fix-pr*

**Comment by athola** [comment]
## PR-Review-Toolkit Comprehensive Analysis

Additional specialized review beyond the initial `/pr-review` findings.

---

### 🔍 Silent Failure Analysis

**Critical (2)**
| Issue | Location | Description |
|-------|----------|-------------|
| SF-1 | `empirical.rs:258-270` | `cluster_sessions` returns empty result silently when data insufficient |
| SF-2 | `behavioral.rs:160-227` | `extract_tool_calls` skips invalid JSON lines without logging |

**Important (4)**
| Issue | Location | Description |
|-------|----------|-------------|
| SF-3 | `comparative.rs:316-329` | `compute_deviation_score` returns None without explanation |
| SF-4 | `install-skrills.sh:456` | Fallback on `claude mcp add` failure may mask CLI issues |
| SF-5 | `behavioral.rs:182-189` | Timestamp defaults to 0 without warning |
| SF-6 | `empirical.rs:655-658` | Default 0.5 confidence used without indication |

**Fix:** Add `tracing::debug\!` logs for early returns and fallback values.

---

### 📋 Test Coverage Analysis

**Critical Gaps (3)**
| Gap | Rating | Description |
|-----|--------|-------------|
| TC-1 | 9/10 | No tests for `extract_tool_calls` JSONL parsing |
| TC-2 | 8/10 | No test verifies actual cluster formation with patterns |
| TC-3 | 8/10 | No test for `compute_effectiveness` with sufficient data |

**Important Gaps (4)**
| Gap | Rating | Description |
|-----|--------|-------------|
| TC-4 | 7/10 | N-gram edge cases (empty sequences, n > length) |
| TC-5 | 7/10 | Deviation score boundary conditions |
| TC-6 | 6/10 | `build_behavioral_patterns` has no tests |
| TC-7 | 5/10 | `compute_feature_distance` is dead code |

**Strengths:** Category inference, baseline expectations, outcome detection, retry patterns well tested.

---

### 🏗️ Type Design Analysis

**Overall Score: 5.5/10**

| Aspect | Rating | Notes |
|--------|--------|-------|
| Encapsulation | 4/10 | All structs use public fields (DTO pattern) |
| Invariant Expression | 5/10 | Constraints documented but not enforced |
| Enum Design | 9/10 | `SkillCategory`, `ToolStatus` well designed |

**Key Issues:**
- `confidence` fields (multiple types) allow values outside [0.0, 1.0]
- `ClusteredBehavior::size` redundant with `session_ids.len()` - can diverge
- String truncation assumes ASCII (Unicode panic risk)

**Recommendations:**
1. Create `Confidence` newtype with clamping constructor
2. Remove redundant `size` field, compute from `session_ids.len()`
3. Use safe truncation for Unicode strings

---

### 📊 Combined Priority Matrix

| Priority | Issues | Action |
|----------|--------|--------|
| **MUST FIX** | B1 (improvement_factor), S4 (Unicode panic), SF-1, SF-2, TC-1 | Fix before merge |
| **SHOULD FIX** | SF-3 to SF-6, TC-2 to TC-6 | Address in this PR |
| **CONSIDER** | Type improvements, G1-G4 suggestions | Future enhancement |

---

### ✅ Positive Observations

1. **Good error messages** in `create_skill_tool` with proper propagation
2. **Preview mode transparency** for empirical generation
3. **Test code** uses `.unwrap()` appropriately (fail fast)
4. **Install script** uses `|| true` for non-critical operations
5. **Clean separation** between modules with clear responsibilities

---

*Generated by `/pr-review-toolkit:review-pr all`*

File: crates/intelligence/src/usage/behavioral.rs [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+                for block in content {
+                    if block.get("type").and_then(|t| t.as_str()) == Some("tool_use") {
+                        if let Some(name) = block.get("name").and_then(|n| n.as_str()) {
+                            let input_summary = block
+                                .get("input")
+                                .map(|i| {
+                                    let s = i.to_string();
+                                    if s.len() > 200 {
+                                        format!("{}...", &s[..197])
+                                    } else {
```
[the remark this answers is not in the document]
Comment (athola) [reply crates/intelligence/src/usage/behavioral.rs]: Fixed in c85c2f2 - Added `safe_truncate()` helper function that respects UTF-8 character boundaries. Applied to both input_summary truncation (line 189) and file path truncation (line 274). Uses `is_char_boundary()` to find valid truncation points.

**Comment by athola** [comment]
### Backlog Issues Created

Following the PR-Review-Toolkit analysis, backlog issues have been created for future improvements:

| Review Item | Issue | Description |
|-------------|-------|-------------|
| Type Improvement | #73 | Add `Confidence` newtype with clamping constructor |
| Type Improvement | #74 | Remove redundant `ClusteredBehavior::size` field |
| SF-3 to SF-6 | #75 | Add tracing for silent early returns and fallbacks |
| TC-2 to TC-7 | #76 | Add tests for behavioral and empirical edge cases |

These were classified as **CONSIDER** / **SHOULD FIX** in the priority matrix and are appropriate for follow-up work after this PR merges.

---
*Backlog issues created from PR #72 review triage*
